internal/push.go

```diff
@@ -0,0 +1,103 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/containerd/containerd/remotes"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	auth "oras.land/oras-go/pkg/auth/docker"
+
+	"oras.land/oras-go/pkg/content"
+	"oras.land/oras-go/pkg/oras"
+)
+
+var push = oras.Push
+
+type PushCommandParams struct {
+	BundleRegistry string
+}
+
+const (
+	customConfigMediaType       = "application/vnd.oci.image.config.v1+json"
+	customTarballLayerMediaType = "application/vnd.oci.image.layer.v1.tar+gzip"
+)
+
+const configContents = `{
+	"mediaType": "application/vnd.oci.image.config.v1+json"
+}`
+
+func RunPush(args []string, params *PushCommandParams) error {
+	ctx := context.Background()
+	memoryStore := content.NewMemoryStore()
+
+	return pushBundle(ctx, memoryStore, params.BundleRegistry, args[0])
+}
+
+func pushBundle(ctx context.Context, memoryStore *content.Memorystore, repository string, path string) error {
+	resolver, err := login(ctx)
+	if err != nil {
+		return err
+	}
+
+	// load the generated bundle and in-memory config.json into OCI descriptors
+	bundle, err := loadBundle(ctx, memoryStore, path)
+	if err != nil {
+		return err
+	}
+	config := loadConfig(memoryStore)
+
+	desriptors := []ocispec.Descriptor{
+		bundle,
+	}
+	pushOpts := configurePushOpts(config)
+	_, err = push(ctx, resolver, repository, memoryStore, desriptors, pushOpts...)
+	if err != nil {
+		return fmt.Errorf("Failed to push bundle to container registry: " + err.Error())
+	}
+
+	return nil
+}
+
+var login = func(ctx context.Context) (remotes.Resolver, error) {
```

Comment:
Consider placing this method in its own file, since the logging in operation might be relevant for other flows.